sql: more redactable details about traces in logs

knz opened the pull request:
Followup to #103034.
Fixes #103319.

Prior to this patch, the logging output for SQL statements / txns that exceeded the tracing threshold would mark the entire SQL statement syntax and the entire trace as redactable.

This commit improves the situation by making the statement syntax and the trace more finely redactable, ensuring that more details are included after redaction.

Note: the redaction is still pretty strong by default, as long as `trace.redactable.enabled` defaults to `false`. See #103884.

Release note (sql change): The logging output emitted when `sql.trace.stmt.enable_threshold` or `sql.trace.txn.enable_threshold` are set is now emited on the SQL_EXEC channel instead of DEV as previously.

Epic: CRDB-27642

knz committed:
sql: more redactable details about traces in logs

Prior to this patch, the logging output for SQL statements / txns that
exceeded the tracing threshold would mark the entire SQL statement
syntax and the entire trace as redactable.

This commit improves the situation by making the statement syntax and
the trace more finely redactable, ensuring that more details are
included after redaction.

Release note (sql change): The logging output emitted when
`sql.trace.stmt.enable_threshold` or `sql.trace.txn.enable_threshold`
are set is now emited on the SQL_EXEC channel instead of DEV as
previously.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/103180)
<!-- Reviewable:end -->

dhartunian reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103180)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @j82w, @knz, @rafiss, and @tbg)

___
*[`pkg/sql/conn_executor_exec.go` line 2943 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103180#-NVeHlUi2Hlh6Px3Kge6:-NVeHlUi2Hlh6Px3Kge7:b-ds53ok) ([raw file](https://github.com/cockroachdb/cockroach/blob/f314420304b93d8523d6928f6aef04f75dab396e/pkg/sql/conn_executor_exec.go#L2943)):*
> ```Go
> ) {
> 	if r == nil {
> 		log.Warning(ctx, "missing trace when threshold tracing was enabled")
> ```

is this a behavior change? we would previously return here.


<!-- Sent from Reviewable.io -->

abarganier reviewed (commented):
Just a couple questions.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103180#-:-NVeKbeFE5ZkuWPrZSPj:bcq77ir)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @j82w, @knz, @rafiss, and @tbg)

___
*[`pkg/sql/conn_executor_exec.go` line 981 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103180#-NVeCplKB8ust-ubDAu6:-NVeCplKB8ust-ubDAu7:b-m955uw) ([raw file](https://github.com/cockroachdb/cockroach/blob/f314420304b93d8523d6928f6aef04f75dab396e/pkg/sql/conn_executor_exec.go#L981)):*
> ```Go
> 				ctx,
> 				rec,
> 				"SQL statement", redactableStmt,
> ```

nit: separate lines for args
___
*[`pkg/sql/conn_executor_exec.go` line 2936 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103180#-NVeD4k-3g475a4Sbl-Y:-NVeD4k-3g475a4Sbl-Z:bnkvvuo) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c9598a9d59433775764c5a78604c2d3a488ad1/pkg/sql/conn_executor_exec.go#L2936)):*
> ```Go
> 	ctx context.Context, r tracingpb.Recording, opName string, threshold, elapsed time.Duration,
> ) {
> 	if elapsed < threshold {
> ```

Just curious, why is this no longer needed? Is it that we're deciding to trust the caller?
___
*[`pkg/sql/conn_executor_exec.go` line 2943 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103180#-NVeHlUi2Hlh6Px3Kge6:-NVeKZRK8YbJJnb1gsD3:b-io5tce) ([raw file](https://github.com/cockroachdb/cockroach/blob/f314420304b93d8523d6928f6aef04f75dab396e/pkg/sql/conn_executor_exec.go#L2943)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

is this a behavior change? we would previously return here.
</blockquote></details>

+1, was wondering about this. 


<!-- Sent from Reviewable.io -->

knz committed:
catalog/lease: safe formatting for descriptorState

Before:
```
descriptorVersionState.incRefCount: ‹101("public") ver=1:1685015076.979754943,0, refcount=1›
```

After:
```
descriptorVersionState.incRefCount: 101(‹"public"›) ver=1:1685015076.979754943,0, refcount=1
```

Release note: None

knz committed:
tracingpb: reveal operation metadata

Before:
```
[batch flow coordinator: ‹{count: 1, duration 1.5s}›]
```

After:
```
[batch flow coordinator: {count: 1, duration 1.5s}]
```

Release note: None

knz committed:
tracing: fix typo in comment

Release note: None

knz committed:
sql: expose engine type in redacted traces

Before:
```
execution starts: ‹distributed› engine
```

After:
```
execution starts: distributed engine
```

Release note: None

knz committed:
execinfrapb: finer grain redactability for ComponentStats

Before:
```
event:‹component:... net_rx:...›
```

After:
```
event:ComponentStats{ID: {‹6e92072b-31ed-4b90-b323-b55d2d2fe0a3› PROCESSOR 0 1}, execution time: 1.5s, batches output: 1, rows output: 1}
```

We can also later consider exposing the FlowID as well, with a
suitable typedef and the `SafeValue` interface.

Release note: None

knz committed:
tracingpb: minor fixup

Release note: None

knz reviewed (commented):
RFAL. I have added a few more commits. Welcoming a review from SQL queries as well.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103180#-:-NWHjlsG20b783bcr5Dy:bhxdoxt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @dhartunian, @j82w, @michae2, and @rafiss)

___
*[`pkg/sql/conn_executor_exec.go` line 981 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103180#-NVeCplKB8ust-ubDAu6:-NWHSrdGDwb3tggYm0i_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f314420304b93d8523d6928f6aef04f75dab396e/pkg/sql/conn_executor_exec.go#L981)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

nit: separate lines for args
</blockquote></details>

Done.
___
*[`pkg/sql/conn_executor_exec.go` line 2936 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103180#-NVeD4k-3g475a4Sbl-Y:-NWHSsoF60ki665hMqeK:b-azuuwb) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c9598a9d59433775764c5a78604c2d3a488ad1/pkg/sql/conn_executor_exec.go#L2936)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

Just curious, why is this no longer needed? Is it that we're deciding to trust the caller?
</blockquote></details>

Yes. It was already the case in one of the 2 callers; I made it true of both. Clarifying in the docstring.
___
*[`pkg/sql/conn_executor_exec.go` line 2943 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103180#-NVeHlUi2Hlh6Px3Kge6:-NWHTkaf7V5PcD1q_spk:b-xtd0f3) ([raw file](https://github.com/cockroachdb/cockroach/blob/f314420304b93d8523d6928f6aef04f75dab396e/pkg/sql/conn_executor_exec.go#L2943)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

+1, was wondering about this. 
</blockquote></details>

Yes, my reasoning is that when a user sets the cluster setting to enable "log upon threshold reached", it's because they want to see _something_ in their exec logs when that threshold has been reached. Even if there's no trace, we can inform them that the threshold has been reached.


<!-- Sent from Reviewable.io -->

rafiss commented on pkg/sql/conn_executor_exec.go:
```diff
@@ -968,17 +968,19 @@ func (ex *connExecutor) execStmtInOpenState(
 			rec := stmtThresholdSpan.FinishAndGetRecording(tracingpb.RecordingVerbose)
 			// NB: This recording does not include the commit for implicit
 			// transactions if the statement didn't auto-commit.
+			redactableStmt := p.FormatAstAsRedactableString(p.stmt.AST,
```
question: should we use `stmt` here? there is logic above to set it:
```
	if isExtendedProtocol {
		stmt = makeStatementFromPrepared(portal.Stmt, queryID)
	} else {
		stmt = makeStatement(parserStmt, queryID)
	}
```

rafiss reviewed (commented):
the catalog and conn_executor changes lgtm

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice!

> Reviewed 3 of 3 files at r2, 1 of 1 files at r3, 2 of 2 files at r4, 2 of 2 files at r5, 1 of 1 files at r6, 1 of 1 files at r7, 5 of 5 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103180#-:-NWK4iWK0RP9WhRSVEmm:b-8gk8uj)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @abarganier, @dhartunian, @j82w, and @knz)

___
*[`pkg/sql/conn_executor_exec.go` line 2943 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103180#-NVeHlUi2Hlh6Px3Kge6:-NWJzjNe9Z2dtSdBFNFE:b5tt9zv) ([raw file](https://github.com/cockroachdb/cockroach/blob/f314420304b93d8523d6928f6aef04f75dab396e/pkg/sql/conn_executor_exec.go#L2943)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Yes, my reasoning is that when a user sets the cluster setting to enable "log upon threshold reached", it's because they want to see _something_ in their exec logs when that threshold has been reached. Even if there's no trace, we can inform them that the threshold has been reached.
</blockquote></details>

The "missing trace" warning will be in a different channel from the trace log message. WDYT about setting r to `"missing trace"` (or something similar) in the nil case?


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103180)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @abarganier, @dhartunian, @j82w, @michae2, and @rafiss)

___
*[`pkg/sql/conn_executor_exec.go` line 2943 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103180#-NVeHlUi2Hlh6Px3Kge6:-NWc-BAK78hORa7sL_mI:bvm4kx7) ([raw file](https://github.com/cockroachdb/cockroach/blob/f314420304b93d8523d6928f6aef04f75dab396e/pkg/sql/conn_executor_exec.go#L2943)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

The "missing trace" warning will be in a different channel from the trace log message. WDYT about setting r to `"missing trace"` (or something similar) in the nil case?
</blockquote></details>

It already does this, see `pkg > util > tracing > tracingpb >  recording.go`, `(r Recording) SafeFormat()`:
```go
  if len(r) == 0 {
    w.SafeString("<empty recording>")
    return
  }
```
___
*[`pkg/sql/conn_executor_exec.go` line 975 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/103180#gh-1205830801:-NWbz_zR6SAy00tV5N9y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa2d863a000daa14c74e9edf1bc1c576640df0b8/pkg/sql/conn_executor_exec.go#L975)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

question: should we use `stmt` here? there is logic above to set it:
```
	if isExtendedProtocol {
		stmt = makeStatementFromPrepared(portal.Stmt, queryID)
	} else {
		stmt = makeStatement(parserStmt, queryID)
	}
```
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

abarganier reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, thank you for this @knz!

I have a couple devil's advocate nits, just double checking to see if we can mark a few extra things as safe. Take it or leave it, I think you're more familiar with the details of these logged vars! 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103180#-:-NWrHTYl7O6U3iDkgOWv:b-2laucn)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @dhartunian, @j82w, @knz, @michae2, and @rafiss)

___
*[`pkg/sql/conn_executor_exec.go` line 2936 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103180#-NVeD4k-3g475a4Sbl-Y:-NWrGPgR2rVMmHAsyJkv:bg2mov3) ([raw file](https://github.com/cockroachdb/cockroach/blob/87c9598a9d59433775764c5a78604c2d3a488ad1/pkg/sql/conn_executor_exec.go#L2936)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Yes. It was already the case in one of the 2 callers; I made it true of both. Clarifying in the docstring.
</blockquote></details>

Gotcha, thanks for adding the docstring! 
___
*[`pkg/sql/catalog/lease/descriptor_version_state.go` line 43 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/103180#-NWrGyzP3sMowiWrD9y6:-NWrGyzP3sMowiWrD9y7:b-kquw1v) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ba34a94b8252949180a2cb461f100baa1ddd92/pkg/sql/catalog/lease/descriptor_version_state.go#L43)):*
> ```Go
> // SafeFormat implements redact.SafeFormatter.
> func (s *storedLease) SafeFormat(w redact.SafePrinter, _ rune) {
> 	w.Printf("ID=%d ver=%d expiration=%s", s.id, s.version, s.expiration)
> ```

nit: can lease IDs, versions, and expirations be marked as safe here? 
___
*[`pkg/sql/catalog/lease/descriptor_version_state.go` line 104 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/103180#-NWrHFIa2iV8obpdcCvj:-NWrHFIa2iV8obpdcCvk:b-qfsik3) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ba34a94b8252949180a2cb461f100baa1ddd92/pkg/sql/catalog/lease/descriptor_version_state.go#L104)):*
> ```Go
> // stringLocked reads mu.refcount and thus needs to have mu held.
> func (s *descriptorVersionState) stringLocked() redact.RedactableString {
> 	return redact.Sprintf("%d(%q) ver=%d:%s, refcount=%d", s.GetID(), s.GetName(), s.GetVersion(), s.mu.expiration, s.mu.refcount)
> ```

nit: ditto here, can any of these be marked safe?



_Code quote:_
```Go
s.GetID(), s.GetName(), s.GetVersion(), s.mu.expiration, s.mu.refcount
```


<!-- Sent from Reviewable.io -->

knz commented:
bors r=abargainier,michae2

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103180)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @abarganier, @dhartunian, @j82w, @michae2, and @rafiss)

___
*[`pkg/sql/catalog/lease/descriptor_version_state.go` line 43 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/103180#-NWrGyzP3sMowiWrD9y6:-NWvCbcj0xwMGKdIMz7N:bi0ez4f) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ba34a94b8252949180a2cb461f100baa1ddd92/pkg/sql/catalog/lease/descriptor_version_state.go#L43)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

nit: can lease IDs, versions, and expirations be marked as safe here? 
</blockquote></details>

They are already. See `init()` in util/log/redact.go.
___
*[`pkg/sql/catalog/lease/descriptor_version_state.go` line 104 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/103180#-NWrHFIa2iV8obpdcCvj:-NWvCkmc-40yCwzEKq2h:b3qjs1u) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ba34a94b8252949180a2cb461f100baa1ddd92/pkg/sql/catalog/lease/descriptor_version_state.go#L104)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

nit: ditto here, can any of these be marked safe?
</blockquote></details>

ditto. (And the name is the only thing not safe)


<!-- Sent from Reviewable.io -->
